packages/messaging/src/models/db-interface.ts

```diff
@@ -18,75 +18,129 @@ import { ErrorFactory } from '@firebase/util';
+  }
+
+  /** Assigns or overwrites the record for the given value. */
+  // tslint:disable-next-line:no-any IndexedDB values are of type "any"
+  put(value: any): Promise<void> {
+    return this.createTransaction(
+      objectStore => objectStore.put(value),
+      'readwrite'
+    );
+  }
 
-    return this.openDbPromise_;
+  /** Deletes record(s) from the objectStore that match the given key. */
+  delete(key: IDBValidKey | IDBKeyRange): Promise<void> {
+    return this.createTransaction(
+      objectStore => objectStore.delete(key),
+      'readwrite'
+    );
   }
 
   /**
    * Close the currently open database.
-   * @return {!Promise} Returns the result of the promise chain.
    */
-  closeDatabase() {
-    return Promise.resolve().then(() => {
-      if (this.openDbPromise_) {
-        return this.openDbPromise_.then(db => {
-          db.close();
-          this.openDbPromise_ = null;
-        });
-      }
-    });
+  async closeDatabase(): Promise<void> {
+    if (this.dbPromise) {
+      const db = await this.dbPromise;
+      db.close();
+      this.dbPromise = null;
+    }
   }
 
   /**
-   * @protected
-   * @param {!IDBDatabase} db
+   * Creates an IndexedDB Transaction and passes its objectStore to the
+   * runRequest function, which runs the database request.
+   *
+   * @return Promise that resolves with the result of the runRequest function
    */
-  onDBUpgrade(db: IDBDatabase, event: IDBVersionChangeEvent) {
-    throw this.errorFactory_.create(ERROR_CODES.SHOULD_BE_INHERITED);
+  private async createTransaction<ReturnType>(
+    runRequest: (objectStore: IDBObjectStore) => IDBRequest,
+    mode?: 'readonly' | 'readwrite'
+  ): Promise<ReturnType> {
+    const db = await this.getDb();
+    const transaction = db.transaction(this.objectStoreName, mode);
+    const request = transaction.objectStore(this.objectStoreName);
+    const result = await promisify<ReturnType>(runRequest(request));
+
+    return new Promise<ReturnType>((resolve, reject) => {
+      transaction.oncomplete = () => {
+        resolve(result);
+      };
+      transaction.onerror = () => {
+        reject(transaction.error);
+      };
+    });
   }
+
+  /** Gets the cached db connection or opens a new one. */
+  private getDb(): Promise<IDBDatabase> {
+    if (!this.dbPromise) {
+      this.dbPromise = new Promise<IDBDatabase>((resolve, reject) => {
+        const request = indexedDB.open(this.dbName, this.dbVersion);
+        request.onsuccess = () => {
+          resolve(request.result);
+        };
+        request.onerror = () => {
+          this.dbPromise = null;
+          reject(request.error);
+        };
+        request.onupgradeneeded = event =>
+          this.onDbUpgrade(request.result, event);
+      });
+    }
+
+    return this.dbPromise;
+  }
+}
+
+/** Promisifies an IDBRequest. Resolves with the IDBRequest's result. */
+function promisify<ReturnType>(request: IDBRequest): Promise<ReturnType> {
```

Comment:
Nit: I've generally seen people use `T` when using generics rather than `ReturnType`.

Reply:
Yeah that makes sense, done.